client/src/main/scala/org/http4s/client/middleware/FollowRedirect.scala

```diff
@@ -23,23 +21,36 @@ object FollowRedirect {
                 fragment = uri.fragment orElse req.uri.fragment
               )
 
-              prepareLoop(
-                req.copy(method = method, uri = nextUri, body = EmptyBody),
-                redirects + 1
-              )
+              // We're following a redirect, so we need to dispose of the last response
+              dispose.flatMap { _ =>
+                prepareLoop(
+                  req.copy(method = method, uri = nextUri, body = EmptyBody),
+                  redirects + 1
+                )
+              }
 
             case _ => Task.now(dr)
           }
         }
 
         resp.status.code match {
-          // We cannot be sure what will happen to the request body so we don't attempt to deal with it
-          case 301 | 302 | 307 | 308 if req.body.isHalt =>
-            dispose.flatMap(_ => doRedirect(req.method))
+          case 301 | 302 =>
+            doRedirect(req.method match {
+              case Method.POST => Method.GET
```

Comment:
In the real world, I expect a GET after a POST on a 302 (this is the classic POST-then-redirect pattern for people who never learned about 303), and I guess I don't have much opinion on 301.  Since we're talking about history, RFC 2616 is also interesting:

> If the 301 status code is received in response to a request other than GET or HEAD, the user agent MUST NOT automatically redirect the request unless it can be confirmed by the user, since this might change the conditions under which the request was issued.
>> Note: When automatically redirecting a POST request after receiving a 301 status code, some existing HTTP/1.0 user agents will erroneously change it into a GET request.

> If the 302 status code is received in response to a request other than GET or HEAD, the user agent MUST NOT automatically redirect the request unless it can be confirmed by the user, since this might change the conditions under which the request was issued.

>> Note: RFC 1945 and RFC 2068 specify that the client is not allowed to change the method on the redirected request.  However, most existing user agent implementations treat 302 as if it were a 303 response, performing a GET on the Location field-value regardless of the original request method. The status codes 303 and 307 have been added for servers that wish to make unambiguously clear which kind of reaction is expected of the client.

There is enough ambiguity in the spec that a strategy with a reasonable default might be good for a minor release, and more sensible defaults implemented here and now.